hatch_pip_compile/plugin.py

```diff
@@ -90,6 +90,14 @@ def get_option_types() -> Dict[str, Any]:
             "pip-compile-install-args": List[str],
         }
 
+    def dependency_hash(self) -> str:
+        """
+        Get the dependency hash
+        """
+        if not self.lockfile_up_to_date:
+            self.sync_dependencies()
+        return super().dependency_hash()
```

Comment:
I don't like calling `lockfile_up_to_date` / `sync_dependencies` here, but I'm not sure there's a way around it

Reply:
https://github.com/pypa/hatch/blob/d3246e957584d292319e7b93301598cdf611e902/src/hatch/cli/application.py#L107-L117

Hatch's workflow:
1. Check current `dependency_hash`
2. Compare it to stored hash
3. If they don't match then run `dependencies_in_sync` / `sync_dependencies` 
4. Update the stored hash with hash from step 1

Reply:
This eliminates the need for `dependencies_in_sync` / `sync_dependencies` to be called after `dependency_hash`

I've opened https://github.com/pypa/hatch/issues/1168 which I think would make all of this a lot cleaner